Blacken tests/functional/

nipunn1313 opened the pull request:
Needed to update black, as test_list.py ran into a bug in black
(that has since been fixed)

nipunn1313 commented:
It appears I need a maintainer approval to run workflows, since I'm a first time contributor. Not sure if this was intentional or not (wasn't the case in the past).

Might need to change this setting https://docs.github.com/en/github/administering-a-repository/managing-repository-settings/disabling-or-limiting-github-actions-for-a-repository#configuring-required-approval-for-workflows-from-public-forks
There are some more permissive settings available.

uranusjr commented:
Could you reduce the scope a bit? It’s not easy to review 47 files and ~10000 LOC in one single PR. Somewhere around 500 would be more reasonable.

nipunn1313 commented:
I limited it to exclude the `test_install` variants (as shown in the .pre-commit-config.yaml)

I'd suggest skipping the detailed review here since it wouldn't have a high utility for the amount of effort.
Splitting the diff file by file also seems like it would have low additional utility relative to the effort given that it's only formatting.

I think it would be more bang for buck to review the changes to `.pre-commit-config.yaml`. The only remaining changes were made by running the linter via `pre-commit run black -a`

pradyunsg commented:
Could you please split out the commit upgrading the black version?

nipunn1313 commented:
Like this? The PR now contains 2 commits.

pfmoore commented:
> I'd suggest skipping the detailed review here since it wouldn't have a high utility for the amount of effort.

As a point of policy, and with no personal comment intended, I'd be cautious of skipping detailed review, especially from someone who is a first time committer. I have seen on other projects, some fairly suspicious changes "hidden" in PRs from newcomers (indeed, that's why github introduced the mechanism of requiring approval to run workflows from first time committers - people were apparently running malicious jobs via other projects' CI 🙁)

Specifically with regard to this PR, I agree with @uranusjr it would be better split into multiple smaller PRs, to make review easier. After all, given that it's only a matter of running black on some files, splitting the PR should be relatively trivial, surely? (The pre-commit change can be made *after* the individual file changes have been done, so that it has a minimal diff).

nipunn1313 commented:
Sure. That’s a valid point. I had assumed some trust which I have not yet
earned.

I am certainly capable of doing the split, but in the end it might be
easier for a trusted committer to just make the change, to avoid having to
review thousands of lines of autoformatted code across tens of prs. It
seems like perhaps having a new committer make this kind of change might
just be counterproductive.

—Nipunn


On Sat, Aug 7, 2021 at 2:04 AM Paul Moore ***@***.***> wrote:

> I'd suggest skipping the detailed review here since it wouldn't have a
> high utility for the amount of effort.
>
> As a point of policy, and with no personal comment intended, I'd be
> cautious of skipping detailed review, especially from someone who is a
> first time committer. I have seen on other projects, some fairly suspicious
> changes "hidden" in PRs from newcomers (indeed, that's why github
> introduced the mechanism of requiring approval to run workflows from first
> time committers - people were apparently running malicious jobs via other
> projects' CI 🙁)
>
> Specifically with regard to this PR, I agree with @uranusjr
> <https://github.com/uranusjr> it would be better split into multiple
> smaller PRs, to make review easier. After all, given that it's only a
> matter of running black on some files, splitting the PR should be
> relatively trivial, surely? (The pre-commit change can be made *after*
> the individual file changes have been done, so that it has a minimal diff).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pypa/pip/pull/10281#issuecomment-894628115>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJ5PI5IF3N2WOOG2NMKNG3T3TZLLANCNFSM5BWZUOMA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

uranusjr commented:
I can’t speak for maintainers, but personally I distrust Black and would read every line from anyone.

pfmoore commented:
> I distrust Black and read every line from anyone in a Blackification PR.

My motives are similar. I actually avoid reviewing PRs that blacken the code, in general, because they tend to make me annoyed and ruin my day. But when I do occasionally review them, I prefer them to be kept small.

> I had assumed some trust which I have not yet earned.

I definitely didn't mean to suggest that I don't trust you - sorry if it sounded like that. My point was a general one, that I don't think we should skip detailed review except in rare cases where a committer who is expert in a particular area is making a relatively straightforward change. And in particular, I think that the discussion that goes on in a review is a valuable way for new contributors and the pip developers to "get to know" each other.

nipunn1313 commented:
Ah - ok thanks for the additional context. I've approached use of black differently and it didn't occur to me to want to double-check the formatter. I've historically treated review as a valuable use of maintainer's time and was coming from a perspective of getting the most value/time for the maintainers here. I now understand that it might look different for pypa/pip community as such a foundational project. Appreciate the extra context and patience you've given and glad to learn.

I was able to split the black into many commits
```
git reset HEAD^
for f in `git status --porcelain | cut -d" " -f3 | grep tests`; do git add $f; git commit -m "Run black on $f"; done
```

There are now 48 separate smaller commits - applying to one file at a time.

nipunn1313 commented:
This is currently "many-commits in one-PR" strategy which git employs https://git-scm.com/docs/SubmittingPatches#separate-commits, though I know that many maintainers like the many-PR strategy. Let me know what is preferred.

nipunn1313 commented:
Also happy to drop the diff entirely if this doesn't seem worth detailing through 46 files.

uranusjr commented:
I think it’s better to use the many PR approach since GitHub doesn’t make it easy to either review or merge one commit at a time (which is more common in the patch-based workflow used by e.g. the Linux kernel).

nipunn1313 commented:
Sounds good. I generated a bunch of PRs via a quick script
```
from subprocess import run
out = run(['git', 'log', '--oneline', 'main..blacken_tests'], capture_output=True).stdout.decode()
hashes = list(reversed([line.split()[0] for line in out.split("\n") if line]))
for i, hsh in enumerate(hashes):
    run(['git', 'checkout', '-b', 'blacken_tests%d' % i, 'main'])
    run(['git', 'cherry-pick', hsh])
```

Will close this one out.

pfmoore commented:
I know I was in favour of more, smaller PRs, but just to note there's no way I'll be reviewing the 40-odd PRs that just landed either. I skimmed a couple, and they looked as OK as anything black generates, but that's as far as I'll go. Apart from anything else, the endless stream of `'` to `"` changes makes it impossible to see what else is going on.

I remain strongly against the idea of mandating that we just refuse to use one of the perfectly valid quote forms Python allows, just because "black makes us". It creates massive churn, makes the black PRs nearly impossible to review, and (IMO) adds nothing in terms of readability.

But I know I've lost that battle, so I'll say no more. Consider this as merely a demonstration of why looking at black-related PRs stresses me out 🙁

nipunn1313 commented:
It sounds like there's maybe some disagreement about whether to use black or not. I don't really have a relevant opinion. I think it's totally fine to decide autoformatting isn't worth it and close all these diffs. I was just trying to help out after looking at the half-finished job in `.pre-commit-config.yaml`.

At my past employer, we used black to specifically avoid having any kind of discussion about formatting or line length. It was built into CI (as it is here) resulting in 0 discussion about formatting - which was a nice positive. To be fair, we didn't really visually look at any of the diffs, and we migrated the entire 10 year old codebase in a handful of commits in one day. Not trying to convince you of anything - just trying to give context as to my confusion on the thread above.

I (like @pfmoore) don't think that it's worth stressing over formatting. Using black (IMO) is supposed to reduce this stress, so if you're finding that it increases such stress, probably we shouldn't be using it (or we're doing something wrong in the process).

pfmoore commented:
@nipunn1313 Apologies, you are indeed getting the fallout of a somewhat unresolved debate over the use of black. Thanks for the work you've done and for caring enough to contribute in the first place. And in particular for being willing to help with the "routine" tasks that sometimes get forgotten about.
